librad/src/id/entity.rs

```diff
@@ -0,0 +1,614 @@
+    #[error("Resolution failed (uri {0})")]
+    ResolutionFailed(String),
+}
+
+#[derive(Debug, Error)]
+pub enum UpdateVerificationError {
+    #[error("Non monotonic revision")]
+    NonMonotonicRevision,
+
+    #[error("Wrong parent hash")]
+    WrongParentHash,
+
+    #[error("Update without previous quorum")]
+    NoPreviousQuorum,
+
+    #[error("Update without current quorum")]
+    NoCurrentQuorum,
+}
+
+#[derive(Debug, Error)]
+pub enum HistoryVerificationError {
+    #[error("Empty history")]
+    EmptyHistory,
+
+    #[error("Error at revsion (rev {revision:?}, err {error:?})")]
+    ErrorAtRevision { revision: u64, error: Error },
+
+    #[error("Update error (rev {revision:?}, err {error:?})")]
+    UpdateError {
+        revision: u64,
+        error: UpdateVerificationError,
+    },
+}
+
+/// A type expressing *who* is signing an `Entity`:
+///
+///  * either a specific user (identified by their URN),
+///  * or the entity itself (with an owned key).
+#[derive(Clone, Debug)]
+pub enum Signatory {
+    User(RadicleUri),
+    OwnedKey,
+}
+
+/// A signature for an `Entity`
+#[derive(Clone, Debug)]
+pub struct EntitySignature {
+    /// Who is producing this signature
+    pub by: Signatory,
+    /// The signature data
+    pub sig: Signature,
+}
+
+/// An URN resolver that turns URNs into `Entity` instances
+/// (`T` is the entity type)
+#[async_trait]
+pub trait Resolver<T> {
+    /// Resolve the given URN and deserialize the target `Entity`
+    async fn resolve(&self, uri: &RadicleUri) -> Result<T, Error>;
+}
+
+/// The base entity definition.
+///
+/// Entities have the following properties:
+///
+/// - They can evolve over time, so they have a sequence of revisions.
+/// - Their identity is stable (it does not change over time), and it is the
+///   hash of their initial revision.
+/// - Each revision contains the hash of the previous revision, which is also
+///   hashed, so that the sequence of revisions is a Merkel tree (actually just
+///   a list).
+/// - They can be signed, either with a key they own, or using a key belonging
+///   to a different entity (the certifier); note that when applying multiple
+///   signatures, signatures are not themselves signed (what is signed is always
+///   only the entity itself).
+/// - Each revision specifies the set of owned keys and trusted certifiers.
+/// - Each revision must be signed by all its owned keys and trusted certifiers.
+/// - Each subsequent revision must be signed by a quorum of the previous keys
+///   and certifiers, to prove that the entity evolution is actually under the
+///   control of its current "owners" (the idea is taken from TUF).
+#[derive(Clone)]
+pub struct Entity<T> {
+    /// The entity name (useful for humans because the hash is unreadable)
+    name: String,
+    /// Entity revision, to be incremented at each entity update
+    revision: u64,
+    /// Entity hash, computed on everything except the signatures and
+    /// (obviously) the hash itself
```

Comment:
I would avoid terms like "obviously" in the documentation. Sometimes things are not so obvious to readers :)
```suggestion
    /// the hash itself
```